token/token.go

```diff
@@ -1,75 +1,146 @@
-	PublicKey() *rsa.PublicKey
+	ParseToken(ctx context.Context, tokenString string) (*TokenClaims, error)
+	PublicKey(kid string) *rsa.PublicKey
+	PublicKeys() []*rsa.PublicKey
 }
 
 type tokenManager struct {
-	publicKey  *rsa.PublicKey
-	privateKey *rsa.PrivateKey
+	publicKeysMap map[string]*rsa.PublicKey
+	publicKeys    []*PublicKey
 }
 
 // NewManager returns a new token Manager for handling tokens
-func NewManager(publicKey *rsa.PublicKey) Manager {
-	return &tokenManager{
-		publicKey: publicKey,
+func NewManager(config configuration) (Manager, error) {
+	// Load public keys from Auth service and add them to the manager
+	tm := &tokenManager{
+		publicKeysMap: map[string]*rsa.PublicKey{},
 	}
-}
 
-// NewManagerWithPrivateKey returns a new token Manager for handling creation of tokens with both private and pulic keys
-func NewManagerWithPrivateKey(privateKey *rsa.PrivateKey) Manager {
-	return &tokenManager{
-		publicKey:  &privateKey.PublicKey,
-		privateKey: privateKey,
-	}
-}
-
-func (mgm tokenManager) Extract(tokenString string) (*account.Identity, error) {
-	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
-		return mgm.publicKey, nil
-	})
+	remoteKeys, err := authtoken.FetchKeys(config.GetKeysEndpoint())
 	if err != nil {
-		return nil, errors.WithStack(err)
+		log.Error(nil, map[string]interface{}{
+			"keys_url": config.GetKeysEndpoint(),
+		}, "unable to load public keys from remote service")
+		return nil, errors.New("unable to load public keys from remote service")
+	}
+	for _, remoteKey := range remoteKeys {
+		tm.publicKeysMap[remoteKey.KeyID] = remoteKey.Key
+		tm.publicKeys = append(tm.publicKeys, &PublicKey{KeyID: remoteKey.KeyID, Key: remoteKey.Key})
+		log.Info(nil, map[string]interface{}{
+			"kid": remoteKey.KeyID,
+		}, "Public key added")
 	}
 
-	if !token.Valid {
-		return nil, errors.New("Token not valid")
+	devModeURL := config.GetKeycloakDevModeURL()
+	if devModeURL != "" {
+		remoteKeys, err = authtoken.FetchKeys(fmt.Sprintf("%s/protocol/openid-connect/certs", devModeURL))
+		if err != nil {
+			log.Error(nil, map[string]interface{}{
+				"keys_url": devModeURL,
+			}, "unable to load public keys from remote service in Dev Mode")
+			return nil, errors.New("unable to load public keys from remote service  in Dev Mode")
+		}
+		for _, remoteKey := range remoteKeys {
+			tm.publicKeysMap[remoteKey.KeyID] = remoteKey.Key
+			tm.publicKeys = append(tm.publicKeys, &PublicKey{KeyID: remoteKey.KeyID, Key: remoteKey.Key})
+			log.Info(nil, map[string]interface{}{
+				"kid": remoteKey.KeyID,
+			}, "Public key added")
+		}
 	}
 
-	claimedUUID := token.Claims.(jwt.MapClaims)["sub"]
-	if claimedUUID == nil {
-		return nil, errors.New("Subject can not be nil")
+	return tm, nil
+}
+
+// NewManagerWithPublicKey returns a new token Manager for handling tokens with the only public key
+func NewManagerWithPublicKey(id string, key *rsa.PublicKey) Manager {
```

Comment:
is this going to be used in dev mode, when the public key for the `test` realm is loaded from the configuration ?

Reply:
This constructor is used by tests only. The tests that don't care about keys from the test realm.